src/main/java/com/microsoft/sqlserver/jdbc/NTLMAuthentication.java

```diff
@@ -0,0 +1,877 @@
+    private final long WINDOWS_EPOCH_DIFF = 11644473600L;
+
+    /**
+     * NTLM Client Context
+     */
+    private class NTLMContext {
+        // domain name to connect to
+        private final String domainName;
+
+        // user credentials
+        private final String userName;
+        private final byte[] passwordHash;
+
+        // upper cased unicode bytes
+        private final byte[] domainBytes;
+        private final byte[] userNameBytes;
+        private final byte[] serverNameBytes;
+        private final byte[] workstationBytes;
+
+        // message authentication code
+        private Mac mac = null;
+
+        // negotiate flags from Challenge msg
+        private long negotiateFlags = 0x00000000;
+
+        // session key calculated from password
+        private byte[] sessionBaseKey = null;
+
+        // Windows FileTime timestamp - number of 100 nanosecond ticks since Windows Epoch time
+        private byte[] timestamp = null;
+
+        // output token
+        private ByteBuffer token = null;
+
+        // target info field from Challenge msg
+        private byte[] targetInfo = null;
+
+        // server challenge from Challenge msg
+        private byte[] serverChallenge = new byte[NTLM_SERVER_CHALLENGE_LENGTH];
+
+        /**
+         * Section 3.1.5.1.2 Client Receives a CHALLENGE_MESSAGE from the Server
+         * 
+         * Save messages for calculating MIC
+         */
+        private byte[] negotiateMsg = null;
+        private byte[] challengeMsg = null;
+
+        /**
+         * Creates an NTLM client context
+         *
+         * @param con
+         * @param serverName
+         * @param domainName
+         * @param workstation
+         * @return NTLM client context
+         * @throws SQLServerException
+         */
+        NTLMContext(final SQLServerConnection con, final String serverName, final String domainName,
+                final String workstation) throws SQLServerException {
+
+            // get server FQDN
+            String serverFqdn = "\0";
+            try {
+                if (null != serverName) {
+                    InetAddress addr = serverName.equals("localhost") ? InetAddress.getByName(workstation)
+                                                                      : InetAddress.getByName(serverName);
+
+                    /*
+                     * Note doc says getCanonicalHostName uses "Best effort method" only, meaning it may not be able to
+                     * return the fqdn
+                     */
+                    serverFqdn = addr.getCanonicalHostName().toUpperCase();
+                }
+            } catch (UnknownHostException e) {
+                MessageFormat form = new MessageFormat(SQLServerException.getErrString("R_ntlmUnknownServer"));
+                Object[] msgArgs = {serverName, e.getMessage()};
+                throw new SQLServerException(form.format(msgArgs), e);
+            }

```

Comment:
There are a few inconsistencies below:

1. Server name, domain name are passed in as parameters, username and password are retrieved from `activeConnectionProperties`.
2. If `serverFqdn` is `null`, we set `serverNameBytes` to `null`, but if `domainName` is null, we set it to `"\0"`.
3. Domain name null check can be done in one line, just like `serverFqdn`.